runtime/runtime.go

```diff
@@ -277,9 +277,14 @@ func (r *interpreterRuntime) parseAndCheckProgram(
 	functions stdlib.StandardLibraryFunctions,
 	options []sema.Option,
 ) (*sema.Checker, error) {
-	program, err := r.parse(code)
+
+	// If the program has not been previously cached, proceed to parse the program.
+	program, err := runtimeInterface.GetCachedProgram(location)
 	if err != nil {
-		return nil, err
+		program, err = r.parse(code)
+		if err != nil {
+			return nil, err
+		}
 	}
```

Comment:
I'd assume `GetCachedProgram` would return `nil`, `nil` for the case where nothing is cached, but also nothing went wrong during the cache lookup. 

Maybe change this to check program instead, and keep the second error return value for actual errors:

```go
if err != nil {
	return nil, err
}

if program == nil {
	program, err = r.parse(code)
	if err != nil {
		return nil, err
	}
}
```

Reply:
Agreed, was kind of playing with what the proper type signature would be for this. Definitely simplifies things 👍